File: BLAS/SRC/dtpsv.f
      SUBROUTINE DTPSV(UPLO,TRANS,DIAG,N,AP,X,INCX)

*  -- Reference BLAS level2 routine --
*  -- Reference BLAS is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      int     INCX,N;
      String    DIAG,TRANS,UPLO;
      // ..
      // .. Array Arguments ..
      double           AP(*),X(*);
      // ..

*  =====================================================================

      // .. Parameters ..
      double           ZERO;
      const     ZERO=0.0D+0;
      // ..
      // .. Local Scalars ..
      double           TEMP;
      int     I,INFO,IX,J,JX,K,KK,KX;
      bool    NOUNIT;
      // ..
      // .. External Functions ..
      bool    LSAME;
      // EXTERNAL LSAME
      // ..
      // .. External Subroutines ..
      // EXTERNAL XERBLA
      // ..

      // Test the input parameters.

      INFO = 0
      if (.NOT.LSAME(UPLO,'U') .AND. .NOT.LSAME(UPLO,'L')) {
          INFO = 1
      } else if (.NOT.LSAME(TRANS,'N') .AND. .NOT.LSAME(TRANS,'T') .AND. .NOT.LSAME(TRANS,'C')) {
          INFO = 2
      } else if (.NOT.LSAME(DIAG,'U') .AND. .NOT.LSAME(DIAG,'N')) {
          INFO = 3
      } else if (N.LT.0) {
          INFO = 4
      } else if (INCX.EQ.0) {
          INFO = 7
      }
      if (INFO.NE.0) {
          CALL XERBLA('DTPSV ',INFO)
          RETURN
      }

      // Quick return if possible.

      IF (N.EQ.0) RETURN

      NOUNIT = LSAME(DIAG,'N')

      // Set up the start point in X if the increment is not unity. This
      // will be  ( N - 1 )*INCX  too small for descending loops.

      if (INCX.LE.0) {
          KX = 1 - (N-1)*INCX
      } else if (INCX.NE.1) {
          KX = 1
      }

      // Start the operations. In this version the elements of AP are
      // accessed sequentially with one pass through AP.

      if (LSAME(TRANS,'N')) {

         // Form  x := inv( A )*x.

          if (LSAME(UPLO,'U')) {
              KK = (N* (N+1))/2
              if (INCX.EQ.1) {
                  DO 20 J = N,1,-1
                      if (X(J).NE.ZERO) {
                          IF (NOUNIT) X(J) = X(J)/AP(KK)
                          TEMP = X(J)
                          K = KK - 1
                          DO 10 I = J - 1,1,-1
                              X(I) = X(I) - TEMP*AP(K)
                              K = K - 1
   10                     CONTINUE
                      }
                      KK = KK - J
   20             CONTINUE
              } else {
                  JX = KX + (N-1)*INCX
                  DO 40 J = N,1,-1
                      if (X(JX).NE.ZERO) {
                          IF (NOUNIT) X(JX) = X(JX)/AP(KK)
                          TEMP = X(JX)
                          IX = JX
                          DO 30 K = KK - 1,KK - J + 1,-1
                              IX = IX - INCX
                              X(IX) = X(IX) - TEMP*AP(K)
   30                     CONTINUE
                      }
                      JX = JX - INCX
                      KK = KK - J
   40             CONTINUE
              }
          } else {
              KK = 1
              if (INCX.EQ.1) {
                  DO 60 J = 1,N
                      if (X(J).NE.ZERO) {
                          IF (NOUNIT) X(J) = X(J)/AP(KK)
                          TEMP = X(J)
                          K = KK + 1
                          DO 50 I = J + 1,N
                              X(I) = X(I) - TEMP*AP(K)
                              K = K + 1
   50                     CONTINUE
                      }
                      KK = KK + (N-J+1)
   60             CONTINUE
              } else {
                  JX = KX
                  DO 80 J = 1,N
                      if (X(JX).NE.ZERO) {
                          IF (NOUNIT) X(JX) = X(JX)/AP(KK)
                          TEMP = X(JX)
                          IX = JX
                          DO 70 K = KK + 1,KK + N - J
                              IX = IX + INCX
                              X(IX) = X(IX) - TEMP*AP(K)
   70                     CONTINUE
                      }
                      JX = JX + INCX
                      KK = KK + (N-J+1)
   80             CONTINUE
              }
          }
      } else {

         // Form  x := inv( A**T )*x.

          if (LSAME(UPLO,'U')) {
              KK = 1
              if (INCX.EQ.1) {
                  DO 100 J = 1,N
                      TEMP = X(J)
                      K = KK
                      DO 90 I = 1,J - 1
                          TEMP = TEMP - AP(K)*X(I)
                          K = K + 1
   90                 CONTINUE
                      IF (NOUNIT) TEMP = TEMP/AP(KK+J-1)
                      X(J) = TEMP
                      KK = KK + J
  100             CONTINUE
              } else {
                  JX = KX
                  DO 120 J = 1,N
                      TEMP = X(JX)
                      IX = KX
                      DO 110 K = KK,KK + J - 2
                          TEMP = TEMP - AP(K)*X(IX)
                          IX = IX + INCX
  110                 CONTINUE
                      IF (NOUNIT) TEMP = TEMP/AP(KK+J-1)
                      X(JX) = TEMP
                      JX = JX + INCX
                      KK = KK + J
  120             CONTINUE
              }
          } else {
              KK = (N* (N+1))/2
              if (INCX.EQ.1) {
                  DO 140 J = N,1,-1
                      TEMP = X(J)
                      K = KK
                      DO 130 I = N,J + 1,-1
                          TEMP = TEMP - AP(K)*X(I)
                          K = K - 1
  130                 CONTINUE
                      IF (NOUNIT) TEMP = TEMP/AP(KK-N+J)
                      X(J) = TEMP
                      KK = KK - (N-J+1)
  140             CONTINUE
              } else {
                  KX = KX + (N-1)*INCX
                  JX = KX
                  DO 160 J = N,1,-1
                      TEMP = X(JX)
                      IX = KX
                      DO 150 K = KK,KK - (N- (J+1)),-1
                          TEMP = TEMP - AP(K)*X(IX)
                          IX = IX - INCX
  150                 CONTINUE
                      IF (NOUNIT) TEMP = TEMP/AP(KK-N+J)
                      X(JX) = TEMP
                      JX = JX - INCX
                      KK = KK - (N-J+1)
  160             CONTINUE
              }
          }
      }

      RETURN

      // End of DTPSV

      }

File: lib/src/blas/dtpsv.dart
import 'package:lapack/src/blas/lsame.dart';
import 'package:lapack/src/blas/xerbla.dart';
import 'package:lapack/src/matrix.dart';

void dtpsv(
  final String UPLO,
  final String TRANS,
  final String DIAG,
  final int N,
  final Array<double> AP_,
  final Array<double> X_,
  final int INCX,
) {
// -- Reference BLAS level2 routine --
// -- Reference BLAS is a software package provided by Univ. of Tennessee,    --
// -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
  final AP = AP_.having();
  final X = X_.having();
  const ZERO = 0.0;

  // Test the input parameters.

  var INFO = 0;
  if (!lsame(UPLO, 'U') && !lsame(UPLO, 'L')) {
    INFO = 1;
  } else if (!lsame(TRANS, 'N') && !lsame(TRANS, 'T') && !lsame(TRANS, 'C')) {
    INFO = 2;
  } else if (!lsame(DIAG, 'U') && !lsame(DIAG, 'N')) {
    INFO = 3;
  } else if (N < 0) {
    INFO = 4;
  } else if (INCX == 0) {
    INFO = 7;
  }
  if (INFO != 0) {
    xerbla('DTPSV ', INFO);
    return;
  }

  // Quick return if possible.

  if (N == 0) return;

  final NOUNIT = lsame(DIAG, 'N');

  // Set up the start point in X if the increment is not unity. This
  // will be  ( N - 1 )*INCX  too small for descending loops.

  var KX = switch (INCX) {
    <= 0 => 1 - (N - 1) * INCX,
    1 => 0,
    _ => 1,
  };

  // Start the operations. In this version the elements of AP are
  // accessed sequentially with one pass through AP.

  if (lsame(TRANS, 'N')) {
    // Form  x := inv( A )*x.

    if (lsame(UPLO, 'U')) {
      var KK = (N * (N + 1)) ~/ 2;
      if (INCX == 1) {
        for (var J = N; J >= 1; J--) {
          if (X[J] != ZERO) {
            if (NOUNIT) X[J] /= AP[KK];
            final TEMP = X[J];
            for (var I = J - 1, K = KK - 1; I >= 1; I--, K--) {
              X[I] -= TEMP * AP[K];
            }
          }
          KK -= J;
        }
      } else {
        var JX = KX + (N - 1) * INCX;
        for (var J = N; J >= 1; J--) {
          if (X[JX] != ZERO) {
            if (NOUNIT) X[JX] /= AP[KK];
            final TEMP = X[JX];
            var IX = JX;
            for (var K = KK - 1; K >= KK - J + 1; K--) {
              IX -= INCX;
              X[IX] -= TEMP * AP[K];
            }
          }
          JX -= INCX;
          KK -= J;
        }
      }
    } else {
      var KK = 1;
      if (INCX == 1) {
        for (var J = 1; J <= N; J++) {
          if (X[J] != ZERO) {
            if (NOUNIT) X[J] /= AP[KK];
            final TEMP = X[J];
            for (var I = J + 1, K = KK + 1; I <= N; I++, K++) {
              X[I] -= TEMP * AP[K];
            }
          }
          KK += N - J + 1;
        }
      } else {
        var JX = KX;
        for (var J = 1; J <= N; J++) {
          if (X[JX] != ZERO) {
            if (NOUNIT) X[JX] /= AP[KK];
            final TEMP = X[JX];
            var IX = JX;
            for (var K = KK + 1; K <= KK + N - J; K++) {
              IX += INCX;
              X[IX] -= TEMP * AP[K];
            }
          }
          JX += INCX;
          KK += N - J + 1;
        }
      }
    }
  } else {
    // Form  x := inv( A**T )*x.

    if (lsame(UPLO, 'U')) {
      var KK = 1;
      if (INCX == 1) {
        for (var J = 1; J <= N; J++) {
          var TEMP = X[J];
          for (var I = 1, K = KK; I <= J - 1; I++, K++) {
            TEMP -= AP[K] * X[I];
          }
          if (NOUNIT) TEMP /= AP[KK + J - 1];
          X[J] = TEMP;
          KK += J;
        }
      } else {
        var JX = KX;
        for (var J = 1; J <= N; J++) {
          var TEMP = X[JX];
          var IX = KX;
          for (var K = KK; K <= KK + J - 2; K++) {
            TEMP -= AP[K] * X[IX];
            IX += INCX;
          }
          if (NOUNIT) TEMP /= AP[KK + J - 1];
          X[JX] = TEMP;
          JX += INCX;
          KK += J;
        }
      }
    } else {
      var KK = (N * (N + 1)) ~/ 2;
      if (INCX == 1) {
        for (var J = N; J >= 1; J--) {
          var TEMP = X[J];
          for (var I = N, K = KK; I >= J + 1; I--, K--) {
            TEMP -= AP[K] * X[I];
          }
          if (NOUNIT) TEMP /= AP[KK - N + J];
          X[J] = TEMP;
          KK -= N - J + 1;
        }
      } else {
        KX += (N - 1) * INCX;
        var JX = KX;
        for (var J = N; J >= 1; J--) {
          var TEMP = X[JX];
          var IX = KX;
          for (var K = KK; K >= KK - (N - (J + 1)); K--) {
            TEMP -= AP[K] * X[IX];
            IX -= INCX;
          }
          if (NOUNIT) TEMP /= AP[KK - N + J];
          X[JX] = TEMP;
          JX -= INCX;
          KK -= N - J + 1;
        }
      }
    }
  }
}
